### DISCUSSION TITLE
How to connect through a corporate proxy?

### DISCUSSION THREAD

**Discussion by NicoKno** [nats-io/nats.net#609] [Q&A] [answered]
Hi,

I'm hosting a NATS-Server in the cloud. When I try to connect from my PC everything works out fine.

When I try to connect from the customers machine I get an error. When I trace the source of the error it seems like the socket connection cannot be established. I suspect that the corporate proxy is getting in the way.

When I go and script a vanilla Socket-Connection I am able to connect the socket to said NATS-Server. But only if I send an HTTP CONNECT with correct credentials to the corp proxy first.

Is there a built in way to do this, or the possibility to provide a Socket "manually"?

Thanks in advance.

**Comment by mtmk** [comment, marked as answer]
You should be able to use WebSockets I'd imagine. As for authentication, would PR #605 solves the issue do you think? maybe you can try that branch if possible.

btw, I'd be interested to see your socket code with http connect. just curious. thanks.

edit: cc @wolfman42 @caleblloyd

**Reply by caleblloyd** [reply]
If you are connecting via WebSocket, then yes #605 should work, set `ClientWebSocketOptions.Proxy` to a [WebProxy](https://learn.microsoft.com/en-us/dotnet/api/system.net.webproxy?view=net-8.0) object I believe

**Reply by NicoKno** [reply]
Looks promising. I'll give it a shot.

The code looks like the following. This worked with our corporate Proxy:
``` csharp
public async static Task<TcpClient> CreateProxied(Uri proxy, Uri destination, string login, string password)
{
	var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);

	await socket.ConnectAsync(destination.Host, destination.Port);

	byte[] connectMessage;
	if (!string.IsNullOrEmpty(login))
	{
		var auth = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{login}:{password}"));
		connectMessage = Encoding.UTF8.GetBytes($"CONNECT {destination.Host}:{destination.Port} HTTP/1.1{Environment.NewLine}Proxy-Authorization: Basic {auth}{Environment.NewLine}{Environment.NewLine}");
	}
	else
	{
		connectMessage = Encoding.UTF8.GetBytes($"CONNECT {destination.Host}:{destination.Port} HTTP/1.1{Environment.NewLine}{Environment.NewLine}");
	}

	socket.Send(connectMessage);

	byte[] receiveBuffer = new byte[1024];
	var received = socket.Receive(receiveBuffer);

	var response = ASCIIEncoding.ASCII.GetString(receiveBuffer, 0, received);

	if (!response.Contains("200 OK"))
	{
		throw new Exception($"Error connecting to proxy server {destination.Host}:{destination.Port}. Response: {response}");
	}

	return new TcpClient
	{
		Client = socket
	};
}
```

Edit: Maybe it would be possible to push a PreRelease?
Edit2: Nevermind, this happens automatically 🤦‍♂️

**Comment by mtmk** [comment]
fyi there is also this
* #647
